### PR TITLE
Add regularizations

### PR DESCRIPTION
**Pull request by jsalbert**
Created a function to go through every Convolutional, Dense and BN layer in the model and apply the specified regularization function.

https://github.com/triagemd/keras-trainer/blob/add_regularizations/keras_trainer/regularizations.py#L5

Based on a discussion here: https://github.com/keras-team/keras/issues/2717

### CODE REVIEW

File: keras_trainer/regularizations.py [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+        regularization_function: Regularization function to include in layers
+        bias_term: If true will add the regularization in the layer bias regularizer
+
+    Returns: Model with regularization function applied
+
+    """
+    # Save weights in case it is a pre-trained model to not lose them afterwards
+    model.save_weights("tmp.h5")
+
+    for layer_index, layer in enumerate(model.layers):
```
Comment (jeremykawahara) [inline keras_trainer/regularizations.py]: Be nice if could select specific layers to add regularisation too

File: keras_trainer/trainer.py [hunk 2]
```diff
@@ -112,7 +115,8 @@ def __init__(self, **options):
             self.train_gen = self.train_generator
 
         if self.num_classes is None and self.top_layers is None:
-            raise ValueError('num_classes must be set to use a custom train_generator with the default fully connected + softmax top_layers')
+            raise ValueError('num_classes must be set to use a custom train_generator with the default fully connected '
+                             '+ softmax top_layers')
```
Comment (jeremykawahara) [inline keras_trainer/trainer.py]: Are the strings split oddly here?

File: keras_trainer/regularizations.py (hunk 1 above)
Comment (jsalbert) [reply keras_trainer/regularizations.py]: Ok added

**Comment by mathmanu** [comment]
How about this:

	# a utility function to add weight decay after the model is defined.
	def add_weight_decay(model, weight_decay):
		if (weight_decay is None) or (weight_decay == 0.0):
			return

		# recursion inside the model
		def add_decay_loss(m, factor):
			if isinstance(m, tf.keras.Model):
				for layer in m.layers:
					add_decay_loss(layer, factor)
			else:
				for param in m.trainable_weights:
					with tf.keras.backend.name_scope('weight_regularizer'):
						regularizer = lambda: tf.keras.regularizers.l2(factor)(param)
						m.add_loss(regularizer)

		# weight decay and l2 regularization differs by a factor of 2
		add_decay_loss(model, weight_decay/2.0)
		return
